### ISSUE TITLE
The implementation of ResidualConnection is not correct?

### ISSUE THREAD

**Issue by imxtx** [hkproj/pytorch-transformer#14]
I think the add operation should be placed before the layer norm:
<img width="329" alt="image" src="https://github.com/hkproj/pytorch-transformer/assets/93981430/e68886f8-4a44-43b7-aaad-7f43fe5430ba">

Correct version:

```python
class ResidualConnection(nn.Module):
    """Implementation of the Residual Connection in the Transformer"""

    def __init__(self, dropout: float) -> None:
        super().__init__()
        self.dropout = nn.Dropout(dropout)
        self.norm = LayerNormalization()

    def forward(self, x, sublayer):
        return self.norm(x + self.dropout(sublayer(x)))
```

**Comment by imxtx** [comment]
Another issue is that there is already a layer norm at the end of each encoder block, so there is no need to put an extra layer norm at the end of the `forward` function in the `Encoder`:

```python
class Encoder(nn.Module):

    def __init__(self, features: int, layers: nn.ModuleList) -> None:
        super().__init__()
        self.layers = layers
        self.norm = LayerNormalization(features)

    def forward(self, x, mask):
        for layer in self.layers:
            x = layer(x, mask)
        return x # instead of return self.norm(x)
```

**Comment by hkproj** [comment]
Hi! My implementation of the Transformer is not 1-to-1 with the paper when it comes to Normalization and Dropout. I have combined the code of multiple implementations, and some of them put the normalization before, while some others put it after. For example in the latest LLaMA 2 model, you will see the Norm blocks before the the Multi-Head-Attention and the Feed-Forward layer. The same goes for the activation functions, there are many variants of the Transformer model which use different activation functions (LLaMA 2 uses the SwiGLU function).

Is it wrong? Is it right? Depends. Different architectures give different results for the same task.

The high-level architecture of my implementation is the same as the paper, but some details may be different.
